File: haskell/state.hs



type State s a = s -> (s,a)

ret :: a -> State s a
ret x = \s -> (s,x)

bind :: State s a -> (a -> State s b) -> State s b
bind m f = \s -> let (s0, a) = m s
                 in f a s0

get :: State s s
get = \s -> (s,s)

put :: s -> State s ()
put v = \s -> (v,())

File: clojure/state.clj



(defn return [a]
	 (fn [s] [s a]))


(defn bind [m f]
  (fn [s] (let [[s2 v] (m s)]
			(f v) s2)))

(defn gets []
  (fn [s] [s s]))

(defn puts [s]
  (fn [_] [s nil]))

(defn mods [f]
  (fn [s] [(f s) nil]))